MR-NIRP Dataset

koutsd opened the pull request:
This PR contains additional features implemented as part of this paper 
https://openaccess.thecvf.com/content/WACV2025W/HAVI/papers/Bouraffa_Deep_Learning-based_rPPG_Models_towards_Automotive_Applications_A_Benchmark_Study_WACVW_2025_paper.pdf

Features:
- Dataset support for the MR-NIRP Car dataset
- ROI Extraction preprocessing step for unsupervised methods using the MediaPipe FaceMesh framework
- Ability to load pretrained weights on CPU (Might not work with models added after June 2024)
- Some minor bug fixes

koutsd commented:
Note: I have no longer access to Nvidia GPUs so I was not able to fully test the toolbox after merging the latest changes.

yahskapar commented on config.py:
```diff
@@ -77,6 +77,9 @@
 _C.TRAIN.DATA.PREPROCESS.CROP_FACE.DETECTION.DO_DYNAMIC_DETECTION = False
 _C.TRAIN.DATA.PREPROCESS.CROP_FACE.DETECTION.DYNAMIC_DETECTION_FREQUENCY = 30
 _C.TRAIN.DATA.PREPROCESS.CROP_FACE.DETECTION.USE_MEDIAN_FACE_BOX = False
+_C.TRAIN.DATA.PREPROCESS.ROI = CN()
```
I'll comment on this more in the rest of the PR, but this won't work given the current pattern of "face detection backends" that we've adopted and currently support Haar Cascade and YOLO5Face. If something based on MediaPipe mesh were to be added, and there's a strong justification with quantitative results to do so, it should be added as a face detection backend rather than creating new config parameters and potentially adding more bloat to this toolbox.

yahskapar commented on config.py:
```diff
@@ -381,7 +386,7 @@ def _update_config_from_file(config, cfg_file):
                 config, os.path.join(os.path.dirname(cfg_file), cfg)
             )
-    print('=> Merging a config file from {}'.format(cfg_file))
+    # print('=> Merging a config file from {}'.format(cfg_file))
```
Don't comment this out - seems unnecessary to change...

yahskapar commented on config.py:
```diff
@@ -356,6 +360,7 @@
 _C.INFERENCE.EVALUATION_WINDOW = CN()
 _C.INFERENCE.EVALUATION_WINDOW.USE_SMALLER_WINDOW = False
 _C.INFERENCE.EVALUATION_WINDOW.WINDOW_SIZE = 10
+_C.INFERENCE.EVALUATION_WINDOW.WINDOW_OVERLAP = 0
```
Why is this added? It doesn't seem to be used anywhere else.

yahskapar commented on config.py:
```diff
@@ -88,10 +91,6 @@
 _C.TRAIN.DATA.PREPROCESS.BIGSMALL.RESIZE.BIG_H = 144
 _C.TRAIN.DATA.PREPROCESS.BIGSMALL.RESIZE.SMALL_W = 9
 _C.TRAIN.DATA.PREPROCESS.BIGSMALL.RESIZE.SMALL_H = 9
-_C.TRAIN.DATA.PREPROCESS.IBVP = CN()
```
Why is anything related to IBVP being removed here and elsewhere?

yahskapar commented on dataset/data_loader/BaseLoader.py:
This feels like too much added code without much justification, though I understand it's related to MediaPipe - can you verify this is even necessary for MR-NIRP? If you use YOLO5Face instead off the main branch of this repo, do you get similar or better results?

koutsd replied:
I will try to check that if I manage to get it set up in a non Nvidia system

yahskapar commented on dataset/data_loader/MRNIRPLoader.py:
```diff
@@ -0,0 +1,282 @@
+        return ppg, frames
+    
+    
+    # def preprocess_dataset(self, data_dirs, config_preprocess, begin=0, end=1):
```
Remove this commented out chunk of code, or justify why it should be left behind.

yahskapar commented on dataset/data_loader/MRNIRPLoader.py:
```diff
@@ -0,0 +1,282 @@
+    def preprocess_dataset_subprocess(self, data_dirs, config_preprocess, i, file_list_dict):
+        """ invoked by preprocess_dataset for multi_process."""        
+        # Skip corrupted frames
+        if data_dirs[i]['index'] == "subject2_garage_small_motion_940":
```
Is there documentation for why this is corrupted, or is this empirically determined?

koutsd replied:
As far as I can tell not officially. This paper  https://arxiv.org/abs/2111.09748 mentions one corrupted file without specifying which. This was the only sample that I found to have corrupted sample which were not readable.

yahskapar commented on dataset/data_loader/MRNIRPLoader.py:
```diff
@@ -0,0 +1,282 @@
+        bvps = self.correct_irregular_sampling(bvps, timestamps, target_fs=self.config_data.FS)
+        bvps, frames = self.match_length(bvps, frames)
+                    
+        # target_length = frames.shape[0]
```
I'd generally remove commented out lines like these and elsewhere, unless there's a good reason to leave them behind.

koutsd replied:
By default the files come compressed and the way to use it "out of the box" would be to use the currently commented lines instead. Decompressing in advance speeds up the process a lot by using the functions named unzipped.

yahskapar commented on dataset/data_loader/MRNIRPLoader.py:
```diff
@@ -0,0 +1,282 @@
+import glob
```
Can you alphabetically sort these imports for readability?

```
import cv2
from dataset.data_loader.BaseLoader import BaseLoader
import glob
import imageio
import io
import numpy as np
import os
import pandas as pd
from scipy.io import loadmat
from tqdm import tqdm
import zipfile
```

This should really be done elsewhere in the repo too, but sadly we've been a bit lax about it.

yahskapar commented on neural_methods/trainer/EfficientPhysTrainer.py:
```diff
@@ -25,28 +25,30 @@ def __init__(self, config, data_loader):
+                print("Loaded Checkpoint:", self.config.INFERENCE.MODEL_PATH)
+
             self.num_train_batches = len(data_loader["train"])
-            self.criterion = torch.nn.MSELoss()
```
Don't change the default loss here.

yahskapar commented on requirements.txt:
Don't change the versions here without justification (e.g., is this something you had to do in your base environment for some reason). Also, let's avoid adding MediaPipe in this same PR - I propose if somehow you try YOLO5Face and it isn't better than MediaPipe, we discuss adding MediaPipe in a completely separate PR dedicated to it.

koutsd replied:
The version changes were to support mediapipe. Generally I can remove all the mediapipe stuff for now

yahskapar commented on neural_methods/trainer/DeepPhysTrainer.py:
```diff
@@ -33,8 +33,12 @@ def __init__(self, config, data_loader):
+                print("Loaded Checkpoint:", self.config.INFERENCE.MODEL_PATH)
+                
             self.num_train_batches = len(data_loader["train"])
-            self.criterion = torch.nn.MSELoss()
```
Don't change the default loss here.

yahskapar commented on neural_methods/trainer/DeepPhysTrainer.py:
Generally speaking, let's avoid the model usage on CPU change in this PR and make a separate one for it as well. As a result, I'd discard these changes in the various trainer files - I'm assuming they generally have something to do with the CPU-based model usage, but there's also some other changes sprinkled here and there that make the changes confusing to review.

yahskapar reviewed (changes requested):
See my comments - notably, I propose removing all changes related to MediaPipe and CPU-based model usage. Instead, consider spinning those sets of changes out into two separate PRs.

yahskapar commented:
> Hello @yahskapar , Reopening the following closed PR #244
> 
> Note: I have no longer access to Nvidia GPUs so I was not able to fully test the toolbox after merging the latest changes.

I just noticed this message, @koutsd - would it be possible for you to test MR-NIRP dataset on your end (if not now, perhaps sometime soon) and see if the minimal addition of the dataset itself would work? It would be a useful dataset to have, though I'm worried some of the other things proposed (MediaPipe, CPU-based model usage) shouldn't be wrapped up in this same PR unless there's a strong justification to do so for the toolbox.

koutsd commented on dataset/data_loader/BaseLoader.py:
```diff
@@ -245,13 +260,16 @@ def preprocess(self, frames, bvps, config_preprocess):
             f_c = frames.copy()
             if data_type == "Raw":
-                data.append(f_c)
+                data.append(f_c / 255.0)
```
This fixed an issue with the visualization notebook where the images were not being displayed. They need to be in 0-1 range.

koutsd commented:
@yahskapar 
I addressed most of your comments. There is a couple of points regarding BaseLoader and MRNIRPLoader that I am not certain of how you would like to handle.

yahskapar commented:
Since PhysDrive has already been onboarded and seems like a reasonably equivalent, more recent dataset, I'll go ahead and close this PR for now. If someone else feels strongly compelled to onboard MR-NIRP, feel free to in a new PR.
